src/lib/pg-bark/_impls/add-appointment.ts

```diff
@@ -0,0 +1,118 @@
+import { Err, Ok, Result } from "@/lib/utilities/result";
+import { CODE } from "@/lib/utilities/bark-code";
+import { PgBarkServiceConfig } from "../pg-bark-service";
+import {
+  dbBegin,
+  dbCommit,
+  dbRelease,
+  dbResultQuery,
+  dbRollback,
+} from "@/lib/data/db-utils";
+import { SQL_QUERY, loadSql } from "../_sql/load-sql";
+import { PoolClient } from "pg";
+
+export async function addAppointment(
+  config: PgBarkServiceConfig,
+  args: {
+    dogId: string;
+    vetId: string;
+  },
+): Promise<
+  Result<
+    { appointmentId: string },
+    | typeof CODE.ERROR_DOG_NOT_FOUND
+    | typeof CODE.ERROR_VET_NOT_FOUND
+    | typeof CODE.ERROR_NOT_PREFERRED_VET
+    | typeof CODE.ERROR_APPOINTMENT_ALREADY_EXISTS
+    | typeof CODE.STORAGE_FAILURE
+  >
+> {
+  const conn = await config.dbPool.connect();
+  const { dogId, vetId } = args;
+  const ctx: _Context = { conn, dogId, vetId };
+  try {
+    await dbBegin(conn);
+    const checkRes = await checkCanSchedule(ctx);
+    if (checkRes !== CODE.OK) {
+      return Err(checkRes);
+    }
+    const { result, error } = await insertAppointment(ctx);
+    if (error !== undefined) {
+      return Err(error);
+    }
+    await dbCommit(conn);
+    return Ok(result);
+  } catch {
+    await dbRollback(conn);
+    return Err(CODE.STORAGE_FAILURE);
+  } finally {
+    await dbRelease(conn);
+  }
+}
+
+type _Context = { conn: PoolClient; dogId: string; vetId: string };
+
+async function checkCanSchedule(args: {
+  conn: PoolClient;
+  dogId: string;
+  vetId: string;
+}): Promise<
+  | typeof CODE.OK
+  | typeof CODE.ERROR_DOG_NOT_FOUND
+  | typeof CODE.ERROR_VET_NOT_FOUND
+  | typeof CODE.ERROR_NOT_PREFERRED_VET
+  | typeof CODE.ERROR_APPOINTMENT_ALREADY_EXISTS
+  | typeof CODE.STORAGE_FAILURE
+> {
+  const { conn, dogId, vetId } = args;
+  const sql = loadSql(SQL_QUERY.SELECT_CAN_SCHEDULE);
+  const { result, error } = await dbResultQuery<{
+    dogExists: boolean;
+    vetExists: boolean;
+    isPreferredVet: boolean;
+    hasExistingAppointment: boolean;
+  }>(conn, sql, [dogId, vetId]);
+  if (error === CODE.DB_QUERY_FAILURE) {
+    return CODE.STORAGE_FAILURE;
+  }
+  if (result.rows.length !== 1) {
+    return CODE.STORAGE_FAILURE;
+  }
+  const { dogExists, vetExists, isPreferredVet, hasExistingAppointment } =
+    result.rows[0];
+  if (!dogExists) {
+    return CODE.ERROR_DOG_NOT_FOUND;
+  }
+  if (!vetExists) {
+    return CODE.ERROR_VET_NOT_FOUND;
+  }
+  if (!isPreferredVet) {
+    return CODE.ERROR_NOT_PREFERRED_VET;
+  }
+  if (hasExistingAppointment) {
+    return CODE.ERROR_APPOINTMENT_ALREADY_EXISTS;
+  }
+  return CODE.OK;
+}
+
+async function insertAppointment(args: {
+  conn: PoolClient;
+  dogId: string;
+  vetId: string;
+}): Promise<Result<{ appointmentId: string }, typeof CODE.STORAGE_FAILURE>> {
+  const { conn, dogId, vetId } = args;
+  const sql = loadSql(SQL_QUERY.INSERT_APPOINTMENT);
+  const { result, error } = await dbResultQuery<{ appointmentId: string }>(
```

Comment:
More and more I feel keeping the SQL in separate files makes the code harder to maintain. The use of dbResultQuery also makes it cumbersome to use. I think each SQL file should be converted into a function that accepts DbContext and args. For example, insertAppointment could be written as shown in Fig1. This way the inputs and outputs are clearly defined.

Fig1: alternative insertAppointment

```ts
async function insertAppointment(
  dbContext: DbContext,
  args: { dogId: string; vetId: string },
): Promise<{ appointmentId: string }> {
  const { dogId, vetId } = args;
  const sql = `
  INSERT INTO calls (dog_id, vet_id, call_outcome, encrypted_opt_out_reason)
  VALUES ($1, $2, 'APPOINTMENT', '')
  RETURNING call_id::text as "appointmentId"  
  `;
  const res = await dbQuery<{ appointmentId: string }>(dbContext, sql, [
    dogId,
    vetId,
  ]);
  return res.rows[0];
}

```

Reply:
Tried this out in alternative-add-appointment.ts. 

It uses insert-appointment.ts and select-appointment-stats.ts

So far so good. The alternative addAppointment is more concise. Let's see how it looks like for something more complicated like create-report.

Reply:
All good now. Refactored everything into lib/bark. PostgreSQL specific code are now in lib/bark/queries. So is that all the value an ORM provides? A replacement to a bunch of query functions? 🤔